pygfx/cameras/_perspective.py

```diff
@@ -1,64 +1,467 @@
+        return self._depth_range
+
+    @depth_range.setter
+    def depth_range(self, value):
+        if value is None:
+            self._depth_range = None
+        elif isinstance(value, (tuple, list)) and len(value) == 2:
+            self._depth_range = float(value[0]), float(value[1])
+        else:
+            raise TypeError("depth_range must be None or a 2-tuple.")
+
+    def _get_near_and_far_plane(self):
+        # Dept range explicitly given?
+        if self._depth_range:
+            return self._depth_range
+
+        # Put 1000 units between the near and far plane, scaled by extent
+        extent = 0.5 * (self._width + self._height)
+        if self.fov > 0:
+            # Scale near plane with the fov to compensate for the fact
+            # that with very small fov you're probably looking at something
+            # in the far distance.
+            f = fov_distance_factor(self.fov)
+            return (extent * f) / 1000, 1000 * extent
+        else:
+            return -500 * extent, 500 * extent
+
+    @property
+    def near(self):
+        """The location of the near clip plane.
+        Use `depth_range` so overload the computed value, if necessary.
+        """
+        near, far = self._get_near_and_far_plane()
+        return near
+
+    @property
+    def far(self):
+        """The location of the far clip plane.
+        Use `depth_range` so overload the computed value, if necessary.
+        """
+        near, far = self._get_near_and_far_plane()
+        return far
+
+    def get_state(self):
+        return {
+            "position": tuple(self.position.to_array()),
+            "rotation": tuple(self.rotation.to_array()),
+            "scale": tuple(self.scale.to_array()),
+            "up": tuple(self.up.to_array()),
+            "fov": self.fov,
+            "width": self.width,
+            "height": self.height,
+            "zoom": self.zoom,
+            "maintain_aspect": self.maintain_aspect,
+            "depth_range": self.depth_range,
+        }
+
+    def set_state(self, state):
+        # Set the more complex props
+        if "position" in state:
+            self.position.set(*state["position"])
+        if "rotation" in state:
+            self.rotation.set(*state["rotation"])
+        if "scale" in state:
+            self.scale.set(*state["scale"])
+        if "up" in state:
+            self.up.set(*state["up"])
+        # Set simple props
+        for key in ("fov", "width", "height", "zoom", "maintain_aspect", "depth_range"):
+            if key in state:
+                setattr(self, key, state[key])
 
     def set_view_size(self, width, height):
         self._view_aspect = width / height
 
     def update_projection_matrix(self):
-        # Get the reference width / height
-        size = 2 * self.near * tan(pi / 180 * 0.5 * self.fov) / self.zoom
-        # Pre-apply the reference aspect ratio
-        width = size * self.aspect**0.5
-        height = size / self.aspect**0.5
```

Comment:
This logic was wrong. It's fixed below. This fix changes how the image is mapped to screen, depending on aspect and window size. This is why some validation examples need their screenshot updated.